### PR TITLE
mDNS Repeater plugin: Enable blocklist parameter in the gui to allow blocking subnets.

### PR DESCRIPTION
**Pull request by Kodehyrden**
This pull request should mitigate the problem described in this forum post:
https://forum.opnsense.org/index.php?topic=44083.0

It will allow the user to add a subnet to be blocked by the mDNS Repeater. The parameter is already available in the program itself. 

This is my very first commit to this repository. Comments are very much welcome. FYI I have also mailed Franz Fabian, maintainer of the mDNS Repeater plugin for OPNsense.

### CODE REVIEW

File: net/mdns-repeater/src/opnsense/mvc/app/controllers/OPNsense/MDNSRepeater/forms/general.xml [hunk 1]
```diff
@@ -17,4 +17,10 @@
     <type>select_multiple</type>
     <help>At least 2 interfaces must be selected. The maximum number of supported interfaces by the daemon is 5.</help>
   </field>
+  <field>
+    <id>mdnsrepeater.bloklist</id>
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/mvc/app/controllers/OPNsense/MDNSRepeater/forms/general.xml:21]: Probably a typo?

File: net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml [hunk 2]
```diff
@@ -16,5 +16,9 @@
       <Required>Y</Required>
       <Multiple>Y</Multiple>
     </interfaces>
+    <blocklist type="TextField">
+      <Default></Default>
+      <Required>N</Required>
+    </interfaces>
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:22]: Typo?

File: net/mdns-repeater/src/opnsense/service/templates/OPNsense/MDNSRepeater/mdnsrepeater [hunk 3]
```diff
@@ -10,6 +10,9 @@ required_files="/var/run/mdns-repeater.CARP_MASTER"
 {% do interface_list.append(physical_interface(i)) %}
 {% endfor %}
 mdns_repeater_interfaces="{{ interface_list | join(' ') }}"
+{% if helpers.exists('OPNsense.MDNSRepeater.bloklist') and OPNsense.MDNSRepeater.bloklist != '' %}
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/service/templates/OPNsense/MDNSRepeater/mdnsrepeater:13]: bloklist or blocklist?

File: net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml [hunk 4]
```diff
@@ -16,5 +16,9 @@
       <Required>Y</Required>
       <Multiple>Y</Multiple>
     </interfaces>
+    <blocklist type="TextField">
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:19]: You could use a NetworkField

https://docs.opnsense.org/development/frontend/models_fieldtypes.html#networkfield

Also is it bloklist or blocklist?

Here examples:
https://github.com/opnsense/plugins/blob/edae632a23c1c35c518539989373187bdd87c520/www/caddy/src/opnsense/mvc/app/models/OPNsense/Caddy/Caddy.xml#L400-L406

https://github.com/opnsense/plugins/blob/edae632a23c1c35c518539989373187bdd87c520/www/caddy/src/opnsense/mvc/app/controllers/OPNsense/Caddy/forms/dialogAccessList.xml#L8-L15

https://github.com/opnsense/plugins/blob/edae632a23c1c35c518539989373187bdd87c520/www/caddy/src/opnsense/service/templates/OPNsense/Caddy/Caddyfile#L86

**Review by Monviech — COMMENTED** [review]
Thank you for your first submission, I have left some general feedback.

**Comment by Monviech** [comment]
https://github.com/geekman/mdns-repeater/releases/tag/1.11

It says here that it supports 16 addresses in a blocklist. Why is it limited to 1 in the PR?

File: net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml [hunk 5]
```diff
@@ -16,5 +16,14 @@
       <Required>Y</Required>
       <Multiple>Y</Multiple>
     </interfaces>
+    <blocklist type="NetworkField">
+      <Default></Default>
+      <Required>N</Required>
+      <ValidationMessage>Please enter one or more valid IPv4 networks or individual addresses in CIDR notation.<ValidationMessage>
+      <NetMaskRequired>Y<NetMaskRequired>
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:23]: Are you testing what you implement here? The xml is wrong. A few `/` are missing.

Please test what you do.

```
root@opn-ce-01:/src/git/opnsense_plugins/net/mdns-repeater # make lint
/src/git/opnsense_plugins/net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:27: parser error : Opening and ending tag mismatch: AsList line 26 and blocklist
    </blocklist>
                ^
/src/git/opnsense_plugins/net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:28: parser error : Opening and ending tag mismatch: AsList line 26 and items
  </items>
          ^
/src/git/opnsense_plugins/net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:29: parser error : Opening and ending tag mismatch: AddressFamily line 25 and model
</model>
        ^
/src/git/opnsense_plugins/net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:30: parser error : Premature end of data in tag AddressFamily line 25

^
*** Error code 1

Stop.
```

**Review by Monviech — COMMENTED** [review]
I want to know why you aren't using the code examples I gave you earlier.

It would allow for multiple networks in the same input field, with the tokenizer separating them cleanly, and commas being automatically added.

Right now I think you expect the user to input their own comma separated list?

Please improve on this concept.

**Comment by Kodehyrden** [comment]
Monviech, first of all, please accept my apologies for the work I have push on you regarding this pull request. I honestly thought it was a lot easier than expected. I have now created a virtual dev environment on my promox. I have installed the modified plugin and it appears now to be working. I did however also realize that changes needs to be made to mens-repeater port. It needs to support black list on the command line as well. I have med the changes, and will create a PR for the ports repository anytime soon. This PR is dependent on acceptance on the ports pull request.

**Comment by Monviech** [comment]
Yeah some things that seem easy can become quite complicated, we're all learning every day. :)

What do you mean you have to do a PR in ports? The ports are mostly synchronized from upstream. Can you explain what exactly you need that is not working right now?

**Comment by Kodehyrden** [comment]
The command line for the mens-repeater need to support blacklists. This is the PR:
Support blacklist on the commandline #208

**Comment by Kodehyrden** [comment]
You're right. Since the code is pulled from https://github.com/kennylevinsen/mdns-repeater/ I will make the pull request here instead. I have deleted my pull request to the ports repository. Oh-my...

**Comment by Monviech** [comment]
Its okay, one step at a time. If you have questions don't be afraid to ask.

**Comment by Kodehyrden** [comment]
It actually turns out that this particular change is not possible to make upstream. Hence the PR for the opnsense ports collection is uphold.

**Comment by Monviech** [comment]
The .in file comes from a different upstream, the FreeBSD ports. They only have a copy on github, the real place to patch that file is in their bugzilla.

Before going that long hard route, I would try if you can script your way around in the jinja template to construct all parameters you need into the existing variable. (if possible)

**Comment by Kodehyrden** [comment]
PR closed in ports collection. Jinja script mods to support blacklist using existing env variable.

File: net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml [hunk 6]
```diff
@@ -1,6 +1,5 @@
 <model>
-  <mount>//OPNsense/MDNSRepeater</mount>
-  <version>1.0.1</version>
+  <mount>//OPNsense/MDNSRepeater</mount>  <version>1.0.1</version>
```
Comment (Monviech) [inline net/mdns-repeater/src/opnsense/mvc/app/models/OPNsense/MDNSRepeater/MDNSRepeater.xml:2]: ?

**Comment by Monviech** [comment]
@Kodehyrden While testing I found some weird things so I made a cleanup of the plugin code and included your change into it. I hope thats okay.

**Comment by Kodehyrden** [comment]
Absolutely - that's perfectly fine. I look forward to seeing your the changes. Should revoke/delete this PR?

Thanks a lot and have a great weekend.

**Comment by Monviech** [comment]
Keep it open for now, we close it when either one gets merged. Thank you for working on this :)

**Comment by Kodehyrden** [comment]
Thank you for all the help and good advice. :-)

**Comment by Monviech** [comment]
Other one got merged. Thanks~

**Pull request closed without merging**